pkg/controller/unreachable/unreachable_controller.go

```diff
@@ -0,0 +1,200 @@
+
+	// remoteClusterAPIClientBuilder is a function pointer to the function that builds a client for the
+	// remote cluster's cluster-api
+	remoteClusterAPIClientBuilder func(string) (client.Client, error)
+}
+
+// Reconcile checks if we can establish an API client connection to the remote cluster and maintains the unreachable condition as a result.
+func (r *ReconcileRemoteMachineSet) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	cd := &hivev1.ClusterDeployment{}
+	err := r.Get(context.TODO(), request.NamespacedName, cd)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request
+		log.WithError(err).Error("error looking up cluster deployment")
+		return reconcile.Result{}, err
+	}
+
+	cdLog := r.logger.WithFields(log.Fields{
+		"clusterDeployment": cd.Name,
+		"namespace":         cd.Namespace,
+	})
+
+	// If the clusterdeployment is deleted, do not reconcile.
+	if cd.DeletionTimestamp != nil {
+		cdLog.Debug("cluster has deletion timestamp")
+		return reconcile.Result{}, nil
+	}
+
+	if !cd.Status.Installed {
+		cdLog.Debug("cluster installation is not complete")
+		return reconcile.Result{}, nil
+	}
+
+	// Check if we're due for rechecking cluster's connectivity
+	cond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.UnreachableCondition)
+	if cond != nil {
+		if !isTimeForConnectivityCheck(cond) {
+			cdLog.WithFields(log.Fields{
+				"sinceLastTransition": time.Since(cond.LastTransitionTime.Time),
+				"sinceLastProbe":      time.Since(cond.LastProbeTime.Time),
+			}).Debug("skipping unreachable check")
+			return reconcile.Result{}, nil
+		}
+	}
+
+	cdLog.Info("checking if cluster is reachable")
+
+	secretName := cd.Status.AdminKubeconfigSecret.Name
+	secretData, err := r.loadSecretData(secretName, cd.Namespace, adminKubeConfigKey)
+	if err != nil {
+		cdLog.WithError(err).Error("unable to load admin kubeconfig")
+		return reconcile.Result{}, err
+	}
+
+	_, err = r.remoteClusterAPIClientBuilder(secretData)
+	if err != nil {
+		cdLog.WithError(err).Info("unable to create remote API client, marking cluster unreachable")
+		cd.Status.Conditions = controllerutils.SetClusterDeploymentCondition(cd.Status.Conditions, hivev1.UnreachableCondition,
+			corev1.ConditionTrue, "ErrorConnectingToCluster", err.Error(), controllerutils.UpdateConditionAlways)
+		err := r.Status().Update(context.TODO(), cd)
+		if err != nil {
+			cdLog.WithError(err).Error("error updating cluster deployment")
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, err
+	}
+
+	// If cluster is reachable, set the "unreachable" condition to false
+	cd.Status.Conditions = controllerutils.SetClusterDeploymentCondition(cd.Status.Conditions, hivev1.UnreachableCondition,
+		corev1.ConditionFalse, "ClusterReachable", err.Error(), controllerutils.UpdateConditionAlways)
+	err = r.Status().Update(context.TODO(), cd)
```

Comment:
If this cluster is going from unreachable to reachable, lets make sure to log that at Info level.

Also per my request in the tests, if a cluster has no unreachable condition, and is still reachable, it should not get an unreachable condition added, that should only happen *if* the cluster has been unreachable at some point.

Reply:
Fixed.